token/program/src/state.rs

```diff
@@ -294,49 +294,162 @@ fn unpack_coption_u64(src: &[u8; 12]) -> Result<COption<u64>, ProgramError> {
+    ) -> Option<T> {
+        if Self::valid_account_data(account_data) {
+            Some(unpack_unchecked_fn(account_data))
+        } else {
+            None
+        }
+    }
+
+    // UNCHECKED FNS
+
+    /// Call after account length has already been verified to unpack the
+    /// account mint
+    fn unpack_account_mint_unchecked(account_data: &[u8]) -> &Pubkey {
+        Self::unpack_pubkey_unchecked(account_data, SPL_TOKEN_ACCOUNT_MINT_OFFSET)
+    }
+
     /// Call after account length has already been verified to unpack the
     /// account owner
     fn unpack_account_owner_unchecked(account_data: &[u8]) -> &Pubkey {
         Self::unpack_pubkey_unchecked(account_data, SPL_TOKEN_ACCOUNT_OWNER_OFFSET)
     }
 
     /// Call after account length has already been verified to unpack the
-    /// account mint
-    fn unpack_account_mint_unchecked(account_data: &[u8]) -> &Pubkey {
-        Self::unpack_pubkey_unchecked(account_data, SPL_TOKEN_ACCOUNT_MINT_OFFSET)
+    /// account amount
+    fn unpack_account_amount_unchecked(account_data: &[u8]) -> u64 {
+        Self::unpack_le_u64_unchecked(account_data, SPL_TOKEN_ACCOUNT_AMOUNT_OFFSET)
     }
 
-    /// Call after account length has already been verified to unpack a Pubkey
-    /// at the specified offset. Panics if `account_data.len()` is less than
-    /// `PUBKEY_BYTES`
-    fn unpack_pubkey_unchecked(account_data: &[u8], offset: usize) -> &Pubkey {
-        bytemuck::from_bytes(&account_data[offset..offset + PUBKEY_BYTES])
+    /// Call after account length has already been verified to unpack the
+    /// account delegate
+    fn unpack_delegate_unchecked(account_data: &[u8]) -> COption<&Pubkey> {
+        Self::unpack_coption_pubkey_unchecked(account_data, SPL_TOKEN_ACCOUNT_DELEGATE_OFFSET)
     }
 
-    /// Unpacks an account's owner from opaque account data.
-    fn unpack_account_owner(account_data: &[u8]) -> Option<&Pubkey> {
-        if Self::valid_account_data(account_data) {
-            Some(Self::unpack_account_owner_unchecked(account_data))
-        } else {
-            None
+    /// Call after account length has already been verified to unpack the
+    /// account state
+    fn unpack_account_state_unchecked(account_data: &[u8]) -> AccountState {
+        AccountState::try_from_primitive(account_data[SPL_TOKEN_ACCOUNT_STATE_OFFSET]).unwrap()
+    }
+
+    /// Call after account length has already been verified to unpack the
+    /// account's is_native field
+    fn unpack_account_is_native_unchecked(account_data: &[u8]) -> COption<u64> {
+        let subslice = &account_data
+            [SPL_TOKEN_ACCOUNT_IS_NATIVE_OFFSET..SPL_TOKEN_ACCOUNT_IS_NATIVE_OFFSET + 12];
+        match subslice[..4] {
+            [0, 0, 0, 0] => COption::None,
+            [1, 0, 0, 0] => {
+                let body: &[u8; 8] = &subslice[4..].try_into().unwrap();
+                COption::Some(u64::from_le_bytes(*body))
+            }
+            _ => panic!(),
         }
```

Comment:
this `COption` unpacking thing  should probably be a generic helper fn somewhere, idk where tho